### DISCUSSION TITLE
How we can enable IPFS replication?

### DISCUSSION THREAD

**Discussion by aledbf** [containerd/nerdctl#966] [Q&A] [answered]
First, thank you for the IPFS example, that just works.

Now, I want to replicate the content to other nodes (the autoscaler could remove a node) and I only see IPFS Cluster on top IPFS. Is that correct? Have you tried something similar to that? Adding IPFS cluster suggests statefulset instead of a daemonset, which defeats the purpose of a variable number of nodes.

Thanks in advance for any suggestion/idea

**Comment by ktock** [comment, marked as answer]
@aledbf Nerdctl + IPFS Cluster on DaemonSet should be possible using something like https://github.com/ktock/stargz-snapshotter/blob/797cf83cabfb6d19240817a665ed6bab78837fc4/script/nerdctl-ipfs-registry-kubernetes-cluster/nerdctl-ipfs-registry.yaml (not well tested. we'll try it deeper and add docs).
Note that 9095 needs to be used as IPFS API port instead of 5001 https://cluster.ipfs.io/documentation/reference/proxy/

**Reply by aledbf** [reply]
@ktock thank you for the quick response and the ipfs-cluster configuration.

The ipfs cluster configuration is similar to what I have with one small difference, I used `replication_factor_max=4`. 
When I do that, the resource utilization is really different.

Does this make sense to you?

``` 
root@server-ws-ephemeral84-m0b3:/home/aledbf# kubectl top pods -n storage
NAME                             CPU(cores)   MEMORY(bytes)   
ipfs-4vnph                       762m         52611Mi         
ipfs-bootstrap-b76c4dc7b-fzc6m   7m           55Mi            
ipfs-dl92k                       21m          1260Mi          
ipfs-fpz7q                       102m         835Mi           
ipfs-hgbb7                       12m          500Mi           
ipfs-jxwxd                       41m          1276Mi          
ipfs-t8xpc                       792m         52521Mi         
``` 

the two pods with high memory are being used for pull/push (and running the replication)

**Comment by aledbf** [comment]
@ktock I find out the reason for the resource utilization issue with `ipfs-cluster`, it was related to the `badgerds` profile. After removing that profile the resource utilization problem is gone

![Screenshot from 2022-04-09 08-22-30](https://user-images.githubusercontent.com/161571/162574174-ab78d7dc-74a0-4139-8a69-cadffe0d1d10.png)

Now, the cluster pulling time is measured in milliseconds now.

Again, thank you for the amazing work with nerdctl.
